### PR TITLE
[SPARK-31803][ML] Make sure instance weight is not negative

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
In the algorithms that support instance weight, add checks to make sure instance weight is not negative.

### Why are the changes needed?
instance weight has to be >= 0.0

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually tested

### CODE REVIEW

**Comment by huaxingao** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/Predictor.scala [hunk 1]
```diff
@@ -79,6 +79,7 @@ private[ml] trait PredictorParams extends Params
 
     dataset.select(col($(labelCol)).cast(DoubleType), w, col($(featuresCol))).rdd.map {
       case Row(label: Double, weight: Double, features: Vector) =>
+        require (weight >= 0.0, "illegal weight value: " + weight + " weight must be >= 0.0")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/Predictor.scala:82]: I was going to wonder if we should require weights to be positive, but I don't think we do now. I guess we tend to catch it when we check whether all weights are 0 (sum = 0).
BTW you can use string interpolation in these error messages.

**Review by srowen — COMMENTED** [review]
Looks generally fine if this consistently checks it in one place, and does so without generating more Spark jobs.

**Comment by zhengruifeng** [comment]
nit: What about adding a `checkNonNegative` (double->double) udf in `ml.functions` and use it in `.ml`? I guess it maybe a little simpler.

**Comment by keypointt** [comment]
Nit: add a util  common method to generate error string? Seems duplicated strings and will need to update each of them if any future changes :)

**Comment by srowen** [comment]
Weird, errors like `org.scalatest.exceptions.TestFailedException: Expected [5.080000000000123,-0.600000000000043] and [-0.96,1.7] to be within 0.001 using absolute tolerance for all elements.` Let me retest

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Merged to master

**Comment by huaxingao** [comment]
Thank you all!

**Pull request closed without merging**